### PR TITLE
Pods watcher initialization and file system watch

### PR DESCRIPTION
**Pull request by sgrebnov**
PRs implements initialization logic for PodsWatcher.

Once changes are detected event handler is called with prev and new pods information.

Next step is to provision vs stop vs update data sources based on prev vs new pods diff.

### CODE REVIEW

File: bin/spiced/src/lib.rs [hunk 1]
```diff
@@ -68,7 +73,10 @@ pub struct Args {
 }
 
 pub async fn run(args: Args) -> Result<()> {
-    let app = App::new(".").context(UnableToConstructSpiceAppSnafu)?;
+
+    let current_dir = env::current_dir().unwrap();
```
Comment (phillipleblanc) [inline bin/spiced/src/lib.rs]: unwrap is not allowed

File: crates/runtime/src/podswatcher.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+                    match event.kind {
+                        EventKind::Create(CreateKind::File) |
+                        EventKind::Remove(RemoveKind::File) |
+                        EventKind::Modify(ModifyKind::Data(_)) => {
+
+                            if let Ok(app) = App::new(root_path.clone()) {
+                                event_handler(PodsWatcherEvent::PodsUpdated(app));
+                            } else {
+                                tracing::debug!("invalid app state detected, ignoring changes.");
```
Comment (phillipleblanc) [inline crates/runtime/src/podswatcher.rs]:
```suggestion
                                tracing::debug!("Invalid app state detected, ignoring changes.");
```

File: crates/runtime/src/podswatcher.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    pub fn watch<F>(&mut self, mut event_handler: F) -> notify::Result<()> 
+    where
+        F: FnMut(PodsWatcherEvent) + 'static + Send,
+    {
+        let root_path = self.root_path.clone();
+
+        let mut watcher = notify::recommended_watcher(move |res: Result<notify::Event, notify::Error>| {
+            match res {
+                Ok(event) => {
+                    tracing::debug!("detected content changes: {:?}", event);
```
Comment (phillipleblanc) [inline crates/runtime/src/podswatcher.rs]:
```suggestion
                    tracing::debug!("Detected content changes: {event:?}");
```

File: crates/runtime/src/podswatcher.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        let root_yaml_files = vec![format!("spicepod.yaml"), format!("spicepod.yml")];
+        for yaml_file in root_yaml_files {
+            let yaml_path = self.root_path.clone().join(&yaml_file);
+            // check that file exists before watching
+            if yaml_path.exists() {
+                watcher.watch(&yaml_path, RecursiveMode::NonRecursive)?;
+            }
+        }
+        
+        // spicepods dir is required to start watcher; create it if it does not exist
```
Comment (phillipleblanc) [inline crates/runtime/src/podswatcher.rs]: We only want to create the spicepods dir if they have dependencies - the watcher shouldn't create it.

File: crates/runtime/src/lib.rs [hunk 5]
```diff
@@ -65,6 +69,27 @@ impl Runtime {
             },
         }
     }
+
+    pub fn start_pods_watcher(&self) -> notify::Result<()> {
+        let mut current_app = self.app.clone();
```
Comment (phillipleblanc) [inline crates/runtime/src/lib.rs]:
```suggestion
        let mut current_app = Arc::clone(&self.app);
```

I find this syntax more explicit.

File: crates/runtime/src/lib.rs [hunk 6]
```diff
@@ -65,6 +69,27 @@ impl Runtime {
...
+
+        let handle_event = move |event: podswatcher::PodsWatcherEvent| {
+            match event {
+                podswatcher::PodsWatcherEvent::PodsUpdated(new_app) => {
+                    tracing::info!("updated pods information: {:?}", new_app);
+                    tracing::info!("previous pods information: {:?}", current_app);
```
Comment (phillipleblanc) [inline crates/runtime/src/lib.rs]: Info level logging is visible to the customers - let's have a high bar for what we log here. This seems like it should be at a debug or trace level, and more concise update for the info level.

File: crates/runtime/src/lib.rs [hunk 7]
```diff
@@ -37,15 +39,17 @@ pub struct Runtime {
     pub app: Arc<app::App>,
     pub config: config::Config,
     pub df: Arc<DataFusion>,
+    pub pods_watcher: RefCell<podswatcher::PodsWatcher>,
```
Comment (phillipleblanc) [inline crates/runtime/src/lib.rs]: This doesn't need to be in a RefCell. Only `start_pods_watcher` needs it to be mutable, and that happens only on initialization. So we can pass a mutable reference to that function directly.

File: bin/spiced/src/lib.rs (hunk 1 above)
Comment (phillipleblanc) [reply bin/spiced/src/lib.rs]: we could default it to "." if the unwrap fails if that is allowed by the file watcher

https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap_or

Comment (sgrebnov) [reply bin/spiced/src/lib.rs]: Updated, thank you!

File: crates/runtime/src/podswatcher.rs (hunk 2 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Fixed, thank you!

File: crates/runtime/src/podswatcher.rs (hunk 3 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Fixed, thank you!

File: crates/runtime/src/podswatcher.rs (hunk 4 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Updated, thank you!

File: crates/runtime/src/lib.rs (hunk 6 above)
Comment (sgrebnov) [reply crates/runtime/src/lib.rs]: Fixed, thank you!

File: crates/runtime/src/lib.rs (hunk 7 above)
Comment (sgrebnov) [reply crates/runtime/src/lib.rs]: Updated, thank you!

File: crates/runtime/src/podswatcher.rs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+fn is_spicepods_modification_event(app_path: impl Into<PathBuf>, event: &notify::Event) -> bool {
+   let root_dir = app_path.into();
+    
+    let spicepod_paths = vec![
+        root_dir.join("spicepods/"),
+        root_dir.join("spicepod.yaml"),
+        root_dir.join("spicepod.yml"),
+    ];
```
Comment (phillipleblanc) [inline crates/runtime/src/podswatcher.rs]: We also need to watch the paths referenced by the dataset/model components.

i.e. This spicepod references a dataset in `datasets/eth.blocks/dataset.yaml`

```yaml
name: app1
version: v1beta1
kind: Spicepod

datasets:
  - ref: datasets/eth.blocks
```

**Review by phillipleblanc — COMMENTED** [review]
Please resolve the conflicts and ensure the CI build/lint is passing, thanks!

File: crates/runtime/src/lib.rs [hunk 9]
```diff
@@ -72,6 +81,27 @@ impl Runtime {
...
         }
     }
+
+    pub fn start_pods_watcher(&mut self) -> notify::Result<()> {
+        let mut current_app = Arc::clone(&self.app);
+
+        let handle_event = move |event: podswatcher::PodsWatcherEvent| {
+            match event {
+                podswatcher::PodsWatcherEvent::PodsUpdated(new_app) => {
+                    tracing::debug!("updated pods information: {:?}", new_app);
```
Comment (mitchdevenport) [inline crates/runtime/src/lib.rs]: Unlike Go, it seems so far that the standard in Rust is to capitalize the first word in logs and errors.

File: crates/runtime/src/podswatcher.rs [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    dirs
+}
+
+fn is_spicepods_modification_event(spicepod_paths: &[PathBuf], event: &notify::Event) -> bool {
+    match event.kind {
+        EventKind::Create(CreateKind::File)
+        | EventKind::Remove(RemoveKind::File)
+        | EventKind::Modify(ModifyKind::Data(_)) => {
+            // iterate through all paths in the event and
+            // check if relative path is within the `spicepods` dir
```
Comment (mitchdevenport) [inline crates/runtime/src/podswatcher.rs]: I would remove this comment, the code is clear enough to be self-documenting.

File: crates/runtime/src/podswatcher.rs [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            match dataset {
+                ComponentOrReference::Reference(reference) => {
+                    let ref_path = root_dir.join(&reference.r#ref);
+                    dirs.push(root_dir.join(ref_path));
+                }
+                ComponentOrReference::Component(_) => { /* ignore component */ }
+            }
+        }
+
+        // watch ref datasets folders
```
Comment (mitchdevenport) [inline crates/runtime/src/podswatcher.rs]: I would remove this comment, the code is clear enough to be self-documenting.

File: crates/runtime/src/podswatcher.rs [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+    // attempt to open root spicepod definition file
+    if let Ok(spicepod) = spicepod::Spicepod::load_definition(&root_dir) {
+        // watch dependencies
+        for dep in spicepod.dependencies {
+            let dep_path = root_dir.join("spicepods").join(dep);
+            dirs.push(dep_path);
+        }
+
+        // watch ref datasets folders
```
Comment (mitchdevenport) [inline crates/runtime/src/podswatcher.rs]: I would remove this comment, the code is clear enough to be self-documenting.

File: crates/runtime/src/podswatcher.rs [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+
+    let mut dirs = vec![
+        // self.root_path.join("spicepods/"),
+        root_dir.join("spicepod.yaml"),
+        root_dir.join("spicepod.yml"),
+    ];
+
+    // attempt to open root spicepod definition file
+    if let Ok(spicepod) = spicepod::Spicepod::load_definition(&root_dir) {
+        // watch dependencies
```
Comment (mitchdevenport) [inline crates/runtime/src/podswatcher.rs]: I would remove this comment, the code is clear enough to be self-documenting.

File: crates/runtime/src/podswatcher.rs [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+fn get_watch_paths(app_path: impl Into<PathBuf>) -> Vec<PathBuf> {
+    let root_dir: PathBuf = app_path.into();
+
+    let mut dirs = vec![
+        // self.root_path.join("spicepods/"),
+        root_dir.join("spicepod.yaml"),
+        root_dir.join("spicepod.yml"),
+    ];
+
+    // attempt to open root spicepod definition file
```
Comment (mitchdevenport) [inline crates/runtime/src/podswatcher.rs]: I would remove this comment, the code is clear enough to be self-documenting.

File: crates/runtime/src/podswatcher.rs [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+        Ok(())
+    }
+}
+
+fn get_watch_paths(app_path: impl Into<PathBuf>) -> Vec<PathBuf> {
+    let root_dir: PathBuf = app_path.into();
+
+    let mut dirs = vec![
+        // self.root_path.join("spicepods/"),
```
Comment (mitchdevenport) [inline crates/runtime/src/podswatcher.rs]: Is this supposed to be commented out?

File: crates/runtime/src/podswatcher.rs (hunk 8 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Implemented 👍

File: crates/runtime/src/lib.rs (hunk 9 above)
Comment (sgrebnov) [reply crates/runtime/src/lib.rs]: Fixed, thank you 👍

File: crates/runtime/src/podswatcher.rs (hunk 11 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Fixed, thank you 👍

File: crates/runtime/src/podswatcher.rs (hunk 13 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Fixed, thank you 👍

File: crates/runtime/src/podswatcher.rs (hunk 15 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Updated, thank you 👍

File: crates/runtime/src/podswatcher.rs (hunk 12 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Fixed, thank you 👍

File: crates/runtime/src/podswatcher.rs (hunk 10 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Fixed, thank you 👍

File: crates/runtime/src/podswatcher.rs (hunk 14 above)
Comment (sgrebnov) [reply crates/runtime/src/podswatcher.rs]: Fixed, thank you 👍

**Review by mitchdevenport — APPROVED** [review]

**Pull request merged**